cmd/agent/main.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/golang/glog"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/spf13/cobra"
+
+	proxyconfig "istio.io/api/proxy/v1/config"
+	"istio.io/pilot/cmd"
+	"istio.io/pilot/proxy"
+	"istio.io/pilot/proxy/envoy"
+	"istio.io/pilot/tools/version"
+)
+
+var (
+	meshconfig string
+	sidecar    proxy.Sidecar
+
+	rootCmd = &cobra.Command{
+		Use:   "agent",
+		Short: "Istio Pilot agent",
+		Long:  "Istio Pilot provides management plane functionality to the Istio service mesh and Istio Mixer.",
+	}
+
+	proxyCmd = &cobra.Command{
+		Use:   "proxy",
+		Short: "Envoy proxy agent",
+		RunE: func(c *cobra.Command, args []string) error {
+			// set values from environment variables
+			if sidecar.IPAddress == "" {
+				sidecar.IPAddress = os.Getenv("POD_IP")
+			}
```

Comment:
Not sure if is wise to encode a common logic like "if no ip address, get Localhost IP". This isn't portable in VMs with multiple interfaces. Either we have a separate main.go for VMs or we expand this such that we provide the interface as an argument. Secondly, we rely on the sidecar having a single IP address. In cases where a VM has multiple interfaces (and hence multiple IP addresses), we need to plumb this all the way to the discovery server. This will quickly get ugly. Lets side step the exact alternatives for now, and simply add a TODO stating that we need to handle scenarios where the host has multiple interfaces.